src/middleware/log.ml

```diff
@@ -339,15 +348,39 @@ let initialize_log
 
   enable := enable_;
   level := level_;
+  let `Initialized = initialized () in
+  ()
+
+module Make (Pclock : Mirage_clock.PCLOCK) = struct
+  let now () = Ptime.to_float_s (Ptime.v (Pclock.now_d_ps ()))
+
+  let initializer_ = lazy begin
+    if !enable then begin
+      Fmt_tty.setup_std_outputs ();
+      Logs.set_level ~all:true (Some !level);
+      Logs.set_reporter (reporter ~now ())
+    end ;
+    `Initialized
+  end
 
-  Lazy.force initializer_
-
+  let set = ref false
 
+  let initialize () =
+    if !set then Logs.debug (fun m -> m "Dream__log.initialize has already been called, \
+                                         ignoring this call.")
+    else begin
+      ( try let `Initialized = initialized () in
+            Format.eprintf "Dream__log.initialize has already been set, check that \
+                            this call is intentional" ;
+        with Logs_are_not_initialized -> () ) ;
+      set := true ;
+      _initialized := Some initializer_
```

Comment:
Can you comment (in the PR) on the need for this two-step initializer (a function, followed by the old lazy initializer)? I suspect it's a way to avoid functorizing too much of the code, i.e. you are trying to inject the initializer from a functor into non-functorized code. But I'd just like to confirm it with you.

Reply:
So I tried to keep the same semantic as before (with `enable` and other values with side-effect and the lazy way to prepare the `reporter`) and put a new indirection. So I kept the `lazy` things as before but we must (see `functoria`/`mirage`) have a _functor entry-point_.

> you are trying to inject the initializer from a functor into non-functorized code.

So yes, this is the final goal to let `functoria` then to apply `dream` with a specific implementation of the posix-clock **and** to not be invasive about what we already have :+1:. But I agree that it's a bit complex to understand especially when the only current case is to apply with `Ptime_clock` - but it's required for MirageOS.